File: Examples/PY/test11.py
def unzip[A, B](l: list[tuple[A, B]]) -> tuple[list[A], list[B]]:
    match l:
        case [h, *t]:
            (h1, h2) = h
            (t1, t2) = unzip(t)
            return ([h1, *t1], [h2, *t2])
        case _:
            return ([], [])

File: Examples/TS/test11.ts
function unzip<A, B>(l: [A, B][]): [A[], B[]] {
    if (l.length === 0) {
        return [[], []];
    } else {
        const h= l[0];
        const t= l.slice(1);
        let [h1, h2] = h;
        let [t1, t2] = unzip(t);
        return [[h1, ...t1], [h2, ...t2]];
    }
}
